### ISSUE TITLE
Linking library by order results in link issue

### ISSUE THREAD

**Issue by tcsunhao** [Open-CMSIS-Pack/devtools#914]
**Describe the bug**
For CMSIS pack, in NXP edgefast blue components, there are lib component like 
![image](https://github.com/Open-CMSIS-Pack/devtools/assets/3381677/0d3ed575-548b-45c6-b591-6c14a24901a7)

If one example depend this component, in the generated cmakelist by cbuild, we'll see
![image](https://github.com/Open-CMSIS-Pack/devtools/assets/3381677/0ad0f2d0-375c-4a49-a0f4-fdd31abc8c2c)

Sometimes, project link fails because of library orders like
![image](https://github.com/Open-CMSIS-Pack/devtools/assets/3381677/a43c6100-7f87-40ba-bdd6-2207b7e22824)

Normally, an archive is searched only once in the order that it is specified on the command line. If a symbol in that archive is needed to resolve an undefined symbol referred to by an object in an archive that appears later on the command line, the linker would not be able to resolve that reference. In our SDK cmake example, we use the --start-group and --end-group command line options to solve this issue, like
![image](https://github.com/Open-CMSIS-Pack/devtools/assets/3381677/73732922-aa8a-47c5-8490-49ccddf21c83)
By grouping the archives, they all be searched repeatedly until all possible references are resolved.

**Expected behavior**
We expect tool can add --start-group and --end-group for libraries to avoid link order issue(circular references). 
Since using this option has a performance cost, we are also ok with some data tips to set certain libraries into --start-group and --end-group.

**Environment (please complete the following information):**

- [tool or library] cbuild 
- Version [e.g. v0.0.3] 1.7
- OS: [e.g. windows/linux/darwin] Windows

Related: https://github.com/Open-CMSIS-Pack/devtools/issues/735

**Comment by ReinhardKeil** [comment]
@tcsunhao a library is a collection of object files.  In your case it seems that each object file is an own library. For a single component I would expect that a single library is sufficient that contains all the object files.   That should be also better for the user.

Is there a technical reason why you are structuring the component into so many different libraries?

As you know the CMSIS-Toolbox is compiler agnostic, we would need to find a way that works on every compiler.

**Comment by brondani** [comment]
@tcsunhao can you please add the following code into your [GCC.\<version>.cmake](https://github.com/Open-CMSIS-Pack/devtools/blob/main/tools/buildmgr/cbuildgen/config/GCC.10.3.1.cmake) file and check whether it solves your problem without impacting performance?
```
# Group libraries for rescanning
set(CMAKE_LINK_GROUP_USING_RESCAN "LINKER:--start-group" "LINKER:--end-group")
set(CMAKE_LINK_GROUP_USING_RESCAN_SUPPORTED TRUE)
set(LIB_FILES "$<LINK_GROUP:RESCAN,${LIB_FILES}>")
```

**Comment by tcsunhao** [comment]
@brondani , I added the 3 lines into my GCC.10.3.1.cmake, unfortunately, it doesn't work. After checking the generated cmakelist.txt, I don't see any change:
![image](https://github.com/Open-CMSIS-Pack/devtools/assets/3381677/8cc95523-a748-4cc1-969d-cfaa2b8aaa59)

**Comment by tcsunhao** [comment]
Hi @ReinhardKeil, not sure about owner's design to this component. I think they are isolated libraries which shall not be merged into a monolithic one. Even we separate each library into each individual component, the tool still needs consider possible link order issue.

**Comment by brondani** [comment]
>After checking the generated cmakelist.txt, I don't see any change

@tcsunhao There is no change in the generated CMakeLists.txt. You should see the impact in `build.ninja` like this:
```
LINK_LIBRARIES = -Wl,--start-group <your libraries> -Wl,--start-group
```

**Comment by tcsunhao** [comment]
@brondani , I open the build.ninja, don't see LINK_LIBRARIES = -Wl,--start-group <your libraries> -Wl,--start-group, just like previous. 

I have deleted all generated tmp/cprj/cbuild yml and try, same issue, it doesn't function.

**Comment by brondani** [comment]
@tcsunhao Could you please provide a reproducer?

**Comment by brondani** [comment]
After testing the previously proposed [patch](https://github.com/Open-CMSIS-Pack/devtools/issues/914#issuecomment-1545360298) with @tcsunhao we realized the [LINK_GROUP](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:LINK_GROUP) feature was first introduced in CMake 3.24.

Another approach compatible with older CMake 3.22 is to simply surround LIB_FILES by start/end group flags in GCC.\<version>.cmake:
```
# Group libraries for rescanning
set(LIB_FILES -Wl,--start-group ${LIB_FILES} -Wl,--end-group)
```
It has fixed the issue concerning circular dependencies and no remarkable performance degradation was observed. For this reason it will be added as default behaviour in the next release.

**Comment by jkrech** [comment]
see: cmsis-toolbox 2.0.0-dev0
